tarofreighter/wallet.go

```diff
@@ -0,0 +1,241 @@
+
+	eligibleCommitments, err := f.cfg.CoinSelector.SelectCommitment(
+		ctx, constraints,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to complete coin "+
+			"selection: %w", err)
+	}
+
+	receiverScriptKey := &receiverAddr.ScriptKey
+	log.Infof("Selected %v possible asset inputs for send to %x",
+		len(eligibleCommitments),
+		receiverScriptKey.SerializeCompressed())
+
+	// We'll take just the first commitment here as we need enough
+	// to complete the send w/o merging inputs.
+	assetInput := eligibleCommitments[0]
+
+	// If the key found for the input UTXO is not from the Taro key family,
+	// something has gone wrong with the DB.
+	internalKey := assetInput.InternalKey
+	if internalKey.Family != taroscript.TaroKeyFamily {
+		return nil, nil, fmt.Errorf("invalid internal key family "+
+			"for selected input: %v %v", internalKey.Family,
+			internalKey.Index)
+	}
+
+	inBip32Derivation, inTrBip32Derivation :=
+		taropsbt.Bip32DerivationFromKeyDesc(
+			internalKey, receiverAddr.ChainParams.HDCoinType,
+		)
+
+	anchorPkScript, anchorMerkleRoot, err := inputAnchorPkScript(assetInput)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot calculate input asset pk "+
+			"script: %w", err)
+	}
+
+	// At this point, we have a valid "coin" to spend in the commitment, so
+	// we'll add the relevant information to the virtual TX's input.
+	//
+	// TODO(roasbeef): still need to add family key to PrevID.
+	pkt := &taropsbt.VPacket{
+		Outputs: make([]*taropsbt.VOutput, 2),
+	}
+	pkt.Input = &taropsbt.VInput{
+		PrevID: asset.PrevID{
+			OutPoint: assetInput.AnchorPoint,
+			ID:       assetInput.Asset.ID(),
+			ScriptKey: asset.ToSerialized(
+				assetInput.Asset.ScriptKey.PubKey,
+			),
+		},
+		Anchor: taropsbt.Anchor{
+			Value:             assetInput.AnchorOutputValue,
+			PkScript:          anchorPkScript,
+			InternalKey:       internalKey.PubKey,
+			MerkleRoot:        anchorMerkleRoot,
+			Bip32Derivation:   inBip32Derivation,
+			TrBip32Derivation: inTrBip32Derivation,
+		},
+		PInput: psbt.PInput{
+			SighashType: txscript.SigHashDefault,
+		},
+	}
+	pkt.SetInputAsset(assetInput.Asset)
+
+	// We'll validate the selected input and commitment. From this we'll
+	// gain the asset that we'll use as an input and info w.r.t if we need
+	// to use an un-spendable zero-value root.
+	inputAsset, fullValue, err := taroscript.IsValidInput(
+		assetInput.Commitment, receiverAddr,
+		*pkt.Input.Asset().ScriptKey.PubKey, *f.cfg.ChainParams,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pkt.Outputs[1] = &taropsbt.VOutput{
+		Amount:            receiverAddr.Amount,
+		IsChange:          false,
+		Interactive:       false,
+		AnchorOutputIndex: 1,
+		ScriptKey:         asset.NewScriptKey(&receiverAddr.ScriptKey),
+	}
+
+	// If we are sending the full value of the input asset, or sending a
+	// collectible, we will need to create a split with un-spendable change.
+	pkt.Outputs[0] = &taropsbt.VOutput{
```

Comment:
So then assumption is that this isn't used for the interactive flow?

Reply:
Yes. I'm not yet sure whether we should just introduce a convention where the _second_ (index 1) output is **always** the recipient and the _first_ (index 0) virtual output is always the change output, even if there is not going to be any change or tombstone in case of an interactive send. Or whether we should always try to dynamically find out which output is for the recipient. I think the latter will be a bit more complicated to follow in the send code, while the former makes it more difficult to use the API because you need to know this convention.

Reply:
Okay, I went with the second option, we now always determine things dynamically, there's no convention other than: **If** there is a change output, it must be at position 0.